packages/lodestar/src/network/peers/peerManager.ts

```diff
@@ -0,0 +1,443 @@
+  SubnetMap,
+  RequestedSubnet,
+  assertPeerRelevance,
+  prioritizePeers,
+} from "./utils";
+
+export interface IPeerManager {
+  start(): void;
+  stop(): void;
+  getConnectedPeerIds(): PeerId[];
+  goodbyeAndDisconnectAllPeers(): Promise<void>;
+  requestAttSubnets(requestedSubnets: RequestedSubnet[]): void;
+  reStatusPeers(peers: PeerId[]): void;
+}
+
+/** heartbeat performs regular updates such as updating reputations and performing discovery requests */
+const HEARTBEAT_INTERVAL_MS = 30 * 1000;
+/** The time in seconds between PING events. We do not send a ping if the other peer has PING'd us */
+const PING_INTERVAL_INBOUND_MS = 14 * 1000; // 1 second faster than lighthouse
+const PING_INTERVAL_OUTBOUND_MS = 17 * 1000;
+/** The time in seconds between re-status's peers. */
+const STATUS_INTERVAL_MS = 5 * 60 * 1000;
+/** Expect a STATUS request from on inbound peer for some time. Afterwards the node does a request */
+const STATUS_INBOUND_GRACE_PERIOD = 15 * 1000;
+
+// TODO:
+// maxPeers and targetPeers should be dynamic on the num of validators connected
+// The Node should compute a recomended value every interval and log a warning
+// to terminal if it deviates significantly from the user's settings
+
+export type PeerManagerOpts = {
+  /** The target number of peers we would like to connect to. */
+  targetPeers: number;
+  /** The maximum number of peers we allow (exceptions for subnet peers) */
+  maxPeers: number;
+};
+
+export type PeerManagerModules = {
+  libp2p: LibP2p;
+  logger: ILogger;
+  metrics: IBeaconMetrics;
+  reqResp: IReqResp;
+  chain: IBeaconChain;
+  config: IBeaconConfig;
+  peerMetadata: Libp2pPeerMetadataStore;
+  peerRpcScores: IPeerRpcScoreStore;
+  networkEventBus: INetworkEventBus;
+};
+
+/**
+ * Tasks:
+ * - Ping peers every `PING_INTERVAL_MS`
+ * - Status peers every `STATUS_INTERVAL_MS`
+ * - Execute discovery query if under target peers
+ * - Execute discovery query if need peers on some subnet: TODO
+ * - Disconnect peers if over target peers
+ */
+export class PeerManager implements IPeerManager {
+  // TODO: Reorg - TEMP
+  private libp2p: LibP2p;
+  private logger: ILogger;
+  private metrics: IBeaconMetrics;
+  private reqResp: IReqResp;
+  private chain: IBeaconChain;
+  private config: IBeaconConfig;
+  private peerMetadata: Libp2pPeerMetadataStore;
+  private peerRpcScores: IPeerRpcScoreStore;
+  private discovery: PeerDiscovery;
+  private networkEventBus: INetworkEventBus;
+
+  /** Map of PeerId -> Time of last PING'd request in ms */
+  private peersToPingOutbound = new PeerMapDelay(PING_INTERVAL_INBOUND_MS);
+  private peersToPingInbound = new PeerMapDelay(PING_INTERVAL_OUTBOUND_MS);
+  /** Map of PeerId -> Time of last STATUS'd request in ms */
+  private peersToStatus = new PeerMapDelay(STATUS_INTERVAL_MS);
+  private opts: PeerManagerOpts;
+  private intervals: NodeJS.Timeout[] = [];
+
+  /** Map of subnets and the slot until they are needed */
+  private subnets = new SubnetMap();
```

Comment:
Yes, I think it's easier since you can compute `activeSubnetIds: number[]` from the `SubnetMap` and then contrast it with the current peer's metadata. It requires no intermediate cached state, which is what the peer duty duration was.

https://github.com/ChainSafe/lodestar/blob/03f356b35be94c51985dc1be87c1489da3d54790/packages/lodestar/src/network/peers/utils/prioritizePeers.ts#L40-L42